### PR TITLE
Sri/ucm 146

### PR DESCRIPTION
**Pull request by srikanth-nalam**
Desired Configuration Used :
{
        "/SDDC-Datacenter/vlcm_cluster1": {
            "config": {
                "profile": {
                    "esx": {
                        "system": {
                            "system_time": {
                                "enabled": true,
                                "protocol": "NTP",
                                "ntp_config": {
                                    "server": [
                                        {
                                            "command": "SERVER",
                                            "hostname": "10.0.0.250"
                                        }
                                    ]
                                }
                            }
                        },
                        "network": {
                            "vmknics": [
                                {
                                    "ip": {
                                        "dhcp": false,
                                        "ipv6": {
                                            "dhcp": false,
                                            "dhcp_dns": false,
                                            "auto_configuration_enabled": true
                                        },
                                        "dhcp_dns": false,
                                        "ipv6_enabled": true
                                    },
                                    "mtu": 9000,
                                    "nic": "vmnic0",
                                    "device": "vmk0",
                                    "enabled": true,
                                    "mac_mode": "PNIC_BASED",
                                    "port_group": "Management Network",
                                    "enabled_services": {
                                        "management": true
                                    },
                                    "port_connection_type": "VSS_PORT_GROUP",
                                    "net_stack_instance_key": "defaultTcpipStack",
                                    "ip_route_default_gateway": "0.0.0.0"
                                }
                            ],
                            "net_stacks": [
                                {
                                    "key": "defaultTcpipStack",
                                    "name": "defaultTcpipStack",
                                    "routing_config": {
                                        "ipv4_default_gateway": "10.0.0.250"
                                    }
                                }
                            ]
                        },
                        "security": {
                            "settings": {
                                "account_lock_failures": 0
                            }
                        },
                        "vmkernel": {
                            "module_settings": [
                                {
                                    "options": "ipv6=1",
                                    "module_name": "tcpip4"
                                },
                                {
                                    "options": "",
                                    "module_name": "i40en"
                                },
                                {
                                    "options": "",
                                    "module_name": "ixgben"
                                },
                                {
                                    "options": "",
                                    "module_name": "smartpqi"
                                },
                                {
                                    "options": "",
                                    "module_name": "irdman"
                                }
                            ]
                        },
                        "network_vss": {
                            "switches": [
                                {
                                    "mtu": 9000,
                                    "name": "vSwitch0",
                                    "bridge": {
                                        "nics": [
                                            "vmnic0"
                                        ],
                                        "link_discovery_protocol": {
                                            "protocol": "CDP",
                                            "operation": "LISTEN"
                                        }
                                    },
                                    "policy": {
                                        "security": {
                                            "mac_changes": false,
                                            "forged_transmits": false,
                                            "allow_promiscuous": false
                                        },
                                        "nic_teaming": {
                                            "policy": "LOADBALANCE_SRCID",
                                            "active_nics": [
                                                "vmnic0"
                                            ],
                                            "standby_nics": [],
                                            "rolling_order": false,
                                            "notify_switches": true,
                                            "link_criteria_beacon": "IGNORE"
                                        }
                                    },
                                    "port_groups": [
                                        {
                                            "name": "Management Network",
                                            "policy": {
                                                "nic_teaming": {
                                                    "policy": "LOADBALANCE_SRCID",
                                                    "active_nics": [
                                                        "vmnic0"
                                                    ],
                                                    "standby_nics": [],
                                                    "rolling_order": false,
                                                    "notify_switches": true,
                                                    "link_criteria_beacon": "IGNORE"
                                                }
                                            },
                                            "vlan_id": 0
                                        },
                                        {
                                            "name": "VM Network",
                                            "vlan_id": 0
                                        }
                                    ]
                                }
                            ]
                        },
                        "authentication": {
                            "user_accounts": [
                                {
                                    "name": "root",
                                    "description": "Administrator",
                                    "password_hash": "$6$ynRHL3AAtGAAWh7b$U9CrYazzCIrabHg8f1twvLl0idtBBrtuvHNESZas7mmOu7JgYn5.WKLfIMq67wB0wb5izTkvlCdyM2mMQw.Sf1"
                                }
                            ]
                        },
                        "advanced_options": {
                            "net": {
                                "tcpip_heap_max": 1024
                            }
                        }
                    }
                },
                "metadata": {
                    "reference_host": {
                        "uuid": "420157bc-3c15-3064-7fdd-7e0de6ec8fae",
                        "build": "Releasebuild-22380479",
                        "patch": "0",
                        "update": "2",
                        "version": "8.0.2"
                    }
                },
                "host-override": {
                    "42013bcb-934c-fb37-60ce-f2545cacbc36": {
                        "esx": {
                            "network": {
                                "net_stacks": [
                                    {
                                        "dns": {
                                            "name_server": [
                                                "10.0.0.250"
                                            ]
                                        },
                                        "key": "defaultTcpipStack",
                                        "name": "defaultTcpipStack",
                                        "host_name": "esxi-7.vrack.vsphere.local",
                                        "routing_config": {
                                            "ipv4_default_gateway": "10.0.0.250"
                                        }
                                    }
                                ]
                            },
                            "authorization": {
                                "permissions": [
                                    {
                                        "is_group": false,
                                        "principal": "svc-vcf-esxi-7",
                                        "access_mode": "Admin"
                                    },
                                    {
                                        "is_group": false,
                                        "principal": "dcui",
                                        "access_mode": "Admin"
                                    },
                                    {
                                        "is_group": false,
                                        "principal": "vpxuser",
                                        "access_mode": "Admin"
                                    },
                                    {
                                        "is_group": false,
                                        "principal": "root",
                                        "access_mode": "Admin"
                                    }
                                ]
                            },
                            "authentication": {
                                "user_accounts": [
                                    {
                                        "name": "root",
                                        "description": "Administrator",
                                        "password_hash": "$6$zMMgP5PHfnJyJEkd$nRsFedHT4JEl2NYk61EVKG5goQoVT2rXvA0WKCqoTBURE3CrSmRh5zvZ8Ptf5aN5tzW99UTIYmw1K.5wns0TX."
                                    },
                                    {
                                        "name": "svc-vcf-esxi-7",
                                        "description": "ESXi User",
                                        "password_hash": "$6$b3znwnCuSCpuF33A$ASmIyrgWJxQZmxxleI7IlgqssppGQ7sVhrJLP6DLH/t8WrqPFtOiuiYmVGy//9UA/srbTzK9vVJN/XLCbFjva."
                                    }
                                ]
                            }
                        }
                    }
                },
                "host-specific": {
                    "42013bcb-934c-fb37-60ce-f2545cacbc36": {
                        "esx": {
                            "network": {
                                "vmknics": [
                                    {
                                        "ip": {
                                            "ipv4_address": "10.0.0.106",
                                            "ipv4_subnet_mask": "255.255.252.0"
                                        },
                                        "device": "vmk0"
                                    }
                                ]
                            }
                        }
                    },
                    "420157bc-3c15-3064-7fdd-7e0de6ec8fae": {
                        "esx": {
                            "network": {
                                "vmknics": [
                                    {
                                        "ip": {
                                            "ipv4_address": "10.0.0.105",
                                            "ipv4_subnet_mask": "255.255.252.0"
                                        },
                                        "device": "vmk0"
                                    }
                                ],
                                "net_stacks": [
                                    {
                                        "key": "defaultTcpipStack",
                                        "host_name": "esxi-6.vrack.vsphere.local"
                                    }
                                ]
                            }
                        }
                    }
                }
            },
            "extended_config": {
                "profile": {
                    "esx": {
                        "services": {
                            "cim_service_enabled": true,
                            "shell_service_enabled": false,
                            "slp_service_enabled": false,
                            "snmp_service_enabled": false,
                            "ssh_service_enabled": true
                        }
                    }
                },
                "metadata": {
                    "reference_host": {
                        "uuid": "420157bc-3c15-3064-7fdd-7e0de6ec8fae"
                    }
                },
                "host-override": {
                    "42013bcb-934c-fb37-60ce-f2545cacbc36": {
                        "esx": {
                            "services": {
                                "cim_service_enabled": false,
                                "ssh_service_enabled": false
                            }
                        }
                    }
                }
            },
            "host_info": {
                "420157bc-3c15-3064-7fdd-7e0de6ec8fae": {
                    "name": "esxi-6.vrack.vsphere.local",
                    "moid": "host-46"
                },
                "42013bcb-934c-fb37-60ce-f2545cacbc36": {
                    "name": "esxi-7.vrack.vsphere.local",
                    "moid": "host-51"
                }
            }
        }
}

    Pre-Check :
salt-call vmware_esxi.precheck cluster_paths=["/SDDC-Datacenter/VLCM_CLUSTER/vlcm_cluster"] -l debug
Pre- Check Execution Response Snapshot :
 

Response In Json Format :
 {
    "/SDDC-Datacenter/vlcm_cluster1": {
        "status": "OK",
        "summary": "Pre-check completed successfully.",
        "hosts": {
            "host-46": {
                "name": "esxi-6.vrack.vsphere.local",
                "summary": "Host is in compliance with desired configuration."
            },
            "host-51": {
                "name": "esxi-7.vrack.vsphere.local",
                "summary": "Host is in compliance with desired configuration."
            }
        },
        "successful_hosts": [
            "host-46",
            "host-51"
        ],
        "failed_hosts": [],
        "skipped_hosts": []
    }
}

Remediate :
salt-call vmware_esxi.remediate cluster_paths=["/SDDC-Datacenter/VLCM_CLUSTER/vlcm_cluster"] -l debug
 Remediate Execution Response Snapshot :
   

Remediate Response Json Format :
{
    "/SDDC-Datacenter/vlcm_cluster1": {
        "hosts": {
            "host-46": {
                "name": "esxi-6.vrack.vsphere.local"
            },
            "host-51": {
                "name": "esxi-7.vrack.vsphere.local"
            }
        },
        "successful_hosts": [
            "host-46",
            "host-51"
        ],
        "failed_hosts": [],
        "skipped_hosts": []
    }
}

### CODE REVIEW

File: src/saltext/vmware/modules/esxi.py [hunk 1]
```diff
@@ -4156,3 +4156,63 @@ def get_reference_schema():
     """
     log.debug("Running vmware_esxi.retrieve_reference_schema")
     return retrieve_reference_schema(Product.ESX)
+
+
+def pre_check(profile=None, cluster_paths=None, desired_state_spec=None, esx_config=None):
```
Comment (dmurphy18) [inline src/saltext/vmware/modules/esxi.py:4161]: As per the meeting today, this should be 'test=True' on remediation

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
test=True

File: src/saltext/vmware/modules/esxi.py (hunk 1 above)
Comment (srikanth-nalam) [reply src/saltext/vmware/modules/esxi.py:4121]: As per the demo given,and in the discussion we have Test=true logic moved to State Module and the changes have been incorporated accordingly

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
just the ret handling, using a dictionary rather than individual 
ret comment

File: src/saltext/vmware/states/esxi.py [hunk 2]
```diff
@@ -1682,6 +1682,117 @@ def vsan_config(
...
+    salt 'minion-id' state.apply your_state_file.sls
+
+    Replace 'minion-id' with the target minion's ID and 'your_state_file.sls' with the path to your state file.
+    """
+    # Initialize the return data structure
+    ret = {
+        "name": name,
+        "result": None,
+        "changes": {},
+        "comment": "Configuration remediated successfully",
```
[the remark this answers is not in the document]
Comment (srikanth-nalam) [reply src/saltext/vmware/states/esxi.py]: Addressed in the latest commit

File: src/saltext/vmware/states/esxi.py [hunk 3]
```diff
@@ -1682,6 +1682,117 @@ def vsan_config(
...
+            )
+
+            if precheck_response["status"] is True:
+                # Pre-check passed, set result to None and proceed with remediation
+                log.debug("Pre-check completed successfully. You can continue with remediation.")
+                ret["result"] = None
+                ret[
+                    "comment"
+                ] = "Pre-check completed successfully. You can continue with remediation."
+                ret["changes"]["pre_check"] = precheck_response
```
[the remark this answers is not in the document]
Comment (srikanth-nalam) [reply src/saltext/vmware/states/esxi.py]: Addressed in the latest commit

**Review by dmurphy18 — APPROVED** [review]

**Pull request merged**